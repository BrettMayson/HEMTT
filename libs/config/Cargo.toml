[package]
name = "hemtt-config"
version = "1.0.0"
edition = "2021"
description = "A config parser library for hemtt"
license = "GPL-2.0"

[lib]
bench = false

[dependencies]
hemtt-common = { path = "../common", version = "1.0.0" }

hemtt-preprocessor = { path = "../preprocessor", version = "1.0.0" }

ariadne = { workspace = true }
byteorder = { workspace = true }
chumsky = "0.9.2"
lsp-types = { workspace = true }
serde = { workspace = true }
vfs = { workspace = true }

[dev-dependencies]
hemtt-preprocessor = { path = "../preprocessor" }
paste = "1.0.14"
