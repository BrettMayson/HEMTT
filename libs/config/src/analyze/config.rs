use std::collections::{HashMap, HashSet};

use hemtt_common::reporting::{Code, Processed};
use hemtt_project::ProjectConfig;

use crate::{Class, Config, Ident, Item, Property, Str, Value};

use super::{
    codes::{
        ce3_duplicate_property::DuplicateProperty, ce7_missing_parent::MissingParent,
        cw1_parent_case::ParentCase, cw2_magwell_missing_magazine::MagwellMissingMagazine,
    },
    Analyze,
};

impl Analyze for Config {
    fn valid(&self, project: Option<&ProjectConfig>) -> bool {
        self.0.iter().all(|p| p.valid(project))
    }

    fn warnings(
        &self,
        project: Option<&ProjectConfig>,
        processed: &Processed,
    ) -> Vec<Box<dyn Code>> {
        let mut warnings = self
            .0
            .iter()
            .flat_map(|p| p.warnings(project, processed))
            .collect::<Vec<_>>();
        let mut defined = HashMap::new();
        warnings.extend(external_parent_case_warn(&self.0, &mut defined));
        if let Some(project) = project {
            warnings.extend(magwell_missing_magazine(project, self));
        }
        warnings
    }

    fn errors(&self, project: Option<&ProjectConfig>, processed: &Processed) -> Vec<Box<dyn Code>> {
        let mut errors = self
            .0
            .iter()
            .flat_map(|p| p.errors(project, processed))
            .collect::<Vec<_>>();
        let mut defined = HashSet::new();
        errors.extend(external_missing_error(&self.0, &mut defined));
        errors.extend(duplicate_properties(&self.0));
        errors
    }
}

fn external_missing_error(
    properties: &[Property],
    defined: &mut HashSet<String>,
) -> Vec<Box<dyn Code>> {
    let mut errors: Vec<Box<dyn Code>> = Vec::new();
    for property in properties {
        if let Property::Class(c) = property {
            match c {
                Class::External { name } => {
                    let name = name.value.to_lowercase();
                    if !defined.contains(&name) {
                        defined.insert(name);
                    }
                }
                Class::Local {
                    name,
                    parent,
                    properties,
                } => {
                    let name = name.value.to_lowercase();
                    if let Some(parent) = parent {
                        let parent = parent.value.to_lowercase();
                        if parent != name && !defined.contains(&parent) {
                            errors.push(Box::new(MissingParent::new(c.clone())));
                        }
                    }
                    defined.insert(name);
                    errors.extend(external_missing_error(properties, defined));
                }
            }
        }
    }
    errors
}

fn external_parent_case_warn(
    properties: &[Property],
    defined: &mut HashMap<String, Class>,
) -> Vec<Box<dyn Code>> {
    let mut warnings: Vec<Box<dyn Code>> = Vec::new();
    for property in properties {
        if let Property::Class(c) = property {
            match c {
                Class::External { name } => {
                    let name = name.value.to_lowercase();
                    defined.entry(name).or_insert_with(|| c.clone());
                }
                Class::Local {
                    name,
                    parent,
                    properties,
                } => {
                    let name_lower = name.value.to_lowercase();
                    if let Some(parent) = parent {
                        let parent_lower = parent.value.to_lowercase();
                        if parent_lower != name_lower {
                            if let Some(parent_class) = defined.get(&parent_lower) {
                                if parent_class.name().value != parent.value {
                                    warnings.push(Box::new(ParentCase::new(
                                        c.clone(),
                                        parent_class.clone(),
                                    )));
                                }
                            }
                        } else if parent.value != name.value {
                            warnings.push(Box::new(ParentCase::new(c.clone(), c.clone())));
                        }
                    }
                    defined.insert(name_lower, c.clone());
                    warnings.extend(external_parent_case_warn(properties, defined));
                }
            }
        }
    }
    warnings
}

fn duplicate_properties(properties: &[Property]) -> Vec<Box<dyn Code>> {
    let mut seen: HashMap<String, Vec<(bool, Ident)>> = HashMap::new();
    duplicate_properties_inner("", properties, &mut seen);
    let mut errors: Vec<Box<dyn Code>> = Vec::new();
    for (_, idents) in seen {
        if idents.len() > 1 && !idents.iter().all(|(class, _)| *class) {
            errors.push(Box::new(DuplicateProperty::new(
                idents.into_iter().map(|(_, i)| i).collect(),
            )));
        }
    }
    errors
}

fn duplicate_properties_inner(
    scope: &str,
    properties: &[Property],
    seen: &mut HashMap<String, Vec<(bool, Ident)>>,
) {
    for property in properties {
        match property {
            Property::Class(Class::Local {
                name, properties, ..
            }) => {
                duplicate_properties_inner(
                    &format!("{}.{}", scope, name.value.to_lowercase()),
                    properties,
                    seen,
                );
                let entry = seen
                    .entry(format!("{}.{}", scope, name.value.to_lowercase()))
                    .or_default();
                entry.push((true, name.clone()));
            }
            Property::Entry { name, .. } | Property::MissingSemicolon(name, _) => {
                let entry = seen
                    .entry(format!("{}.{}", scope, name.value.to_lowercase()))
                    .or_default();
                entry.push((false, name.clone()));
            }
            _ => (),
        }
    }
}

fn magwell_missing_magazine(project: &ProjectConfig, config: &Config) -> Vec<Box<dyn Code>> {
    let mut warnings: Vec<Box<dyn Code>> = Vec::new();
    let mut classes = Vec::new();
    let Some(Property::Class(Class::Local {
        properties: magwells,
        ..
    })) = config
        .0
        .iter()
        .find(|p| p.name().value.to_lowercase() == "cfgmagazinewells")
    else {
        return warnings;
    };
    let Some(Property::Class(Class::Local {
        properties: magazines,
        ..
    })) = config
        .0
        .iter()
        .find(|p| p.name().value.to_lowercase() == "cfgmagazines")
    else {
        return warnings;
    };
    for property in magazines {
        if let Property::Class(Class::Local { name, .. }) = property {
            classes.push(name);
        }
    }
    for magwell in magwells {
        let Property::Class(Class::Local {
            properties: addons, ..
        }) = magwell
        else {
            continue;
        };
        for addon in addons {
            let Property::Entry {
                name,
                value: Value::Array(magazines),
                ..
            } = addon
            else {
                continue;
            };
            for mag in &magazines.items {
                let Item::Str(Str { value, span }) = mag else {
                    continue;
                };
                if !value
                    .to_lowercase()
                    .starts_with(&project.prefix().to_lowercase())
                {
                    continue;
                }
                if !classes.iter().any(|c| c.value == *value) {
                    warnings.push(Box::new(MagwellMissingMagazine::new(
                        name.clone(),
                        value.clone(),
                        span.clone(),
                    )));
                }
            }
        }
    }
    warnings
}
