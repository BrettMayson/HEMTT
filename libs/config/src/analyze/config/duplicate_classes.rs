use std::{collections::HashMap, sync::Arc};

use hemtt_workspace::reporting::{Code, Processed};

use crate::{analyze::codes::ce9_duplicate_classes::DuplicateClasses, Class, Property};

pub fn error(properties: &[Property], processed: &Processed) -> Vec<Arc<dyn Code>> {
    let mut defined: HashMap<String, Vec<Class>> = HashMap::new();
    let mut errors = Vec::new();
    for property in properties {
        if let Property::Class(c) = property {
            match c {
                Class::Root { properties } => {
                    errors.extend(error(properties, processed));
                }
                Class::External { name } => {
                    defined
                        .entry(name.value.to_lowercase())
                        .or_default()
                        .push(c.clone());
                }
                Class::Local {
                    name,
                    parent: _,
                    properties,
                } => {
                    errors.extend(error(properties, processed));
                    defined
                        .entry(name.value.to_lowercase())
                        .or_default()
                        .push(c.clone());
                }
            }
        }
    }
    errors.extend(defined.into_iter().filter_map(|(_, classes)| {
        if classes.len() > 1 {
            Some(Arc::new(DuplicateClasses::new(classes, processed)) as Arc<dyn Code>)
        } else {
            None
        }
    }));
    errors
}
