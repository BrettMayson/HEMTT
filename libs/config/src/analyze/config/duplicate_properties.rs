use std::{collections::HashMap, sync::Arc};

use hemtt_workspace::reporting::{Code, Processed};

use crate::{analyze::codes::ce3_duplicate_property::DuplicateProperty, Class, Ident, Property};

pub fn duplicate_properties(properties: &[Property], processed: &Processed) -> Vec<Arc<dyn Code>> {
    let mut seen: HashMap<String, Vec<(bool, Ident)>> = HashMap::new();
    duplicate_properties_inner("", properties, &mut seen);
    let mut errors: Vec<Arc<dyn Code>> = Vec::new();
    for (_, idents) in seen {
        if idents.len() > 1 && !idents.iter().all(|(class, _)| *class) {
            errors.push(Arc::new(DuplicateProperty::new(
                idents.into_iter().map(|(_, i)| i).collect(),
                processed,
            )));
        }
    }
    errors
}

fn duplicate_properties_inner(
    scope: &str,
    properties: &[Property],
    seen: &mut HashMap<String, Vec<(bool, Ident)>>,
) {
    for property in properties {
        match property {
            Property::Class(Class::Local {
                name, properties, ..
            }) => {
                duplicate_properties_inner(
                    &format!("{}.{}", scope, name.value.to_lowercase()),
                    properties,
                    seen,
                );
                let entry = seen
                    .entry(format!("{}.{}", scope, name.value.to_lowercase()))
                    .or_default();
                entry.push((true, name.clone()));
            }
            Property::Entry { name, .. } | Property::MissingSemicolon(name, _) => {
                let entry = seen
                    .entry(format!("{}.{}", scope, name.value.to_lowercase()))
                    .or_default();
                entry.push((false, name.clone()));
            }
            _ => (),
        }
    }
}
