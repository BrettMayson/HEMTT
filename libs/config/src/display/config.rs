use crate::Config;

impl std::fmt::Display for Config {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for property in &self.0 {
            write!(f, "{property}")?;
        }
        Ok(())
    }
}
