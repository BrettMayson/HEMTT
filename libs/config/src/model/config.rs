use hemtt_common::version::Version;

use crate::{analyze::CfgPatch, Class, Number, Property, Value};

#[derive(Debug, PartialEq)]
/// A config file
pub struct Config(pub Vec<Property>);

impl Config {
    #[must_use]
    pub fn to_class(&self) -> Class {
        Class::Root {
            properties: self.0.clone(),
        }
    }
}

impl Config {
    #[must_use]
    /// Get the patches
    pub fn get_patches(&self) -> Vec<CfgPatch> {
        let mut patches = Vec::new();
        for property in &self.0 {
            if let Property::Class(Class::Local {
                name, properties, ..
            }) = property
            {
                if name.as_str() == "CfgPatches" {
                    for patch in properties {
                        if let Property::Class(Class::Local {
                            name, properties, ..
                        }) = patch
                        {
                            let mut required_version = Version::new(0, 0, 0, None);
                            for property in properties {
                                if let Property::Entry { name, value, .. } = property {
                                    if name.as_str() == "requiredVersion" {
                                        if let Value::Number(Number::Float32 { value, .. }) = value
                                        {
                                            required_version = Version::from(*value);
                                        }
                                    }
                                }
                            }
                            patches.push(CfgPatch::new(name.clone(), required_version));
                        }
                    }
                }
            }
        }
        patches
    }
}
