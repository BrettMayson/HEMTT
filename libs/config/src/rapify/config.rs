use std::io::Cursor;

use byteorder::{LittleEndian, WriteBytesExt};

use crate::{Class, Config};

use super::Rapify;

impl Rapify for Config {
    fn rapify<O: std::io::Write>(
        &self,
        output: &mut O,
        _offset: usize,
    ) -> Result<usize, std::io::Error> {
        output.write_all(b"\0raP")?;
        output.write_all(b"\0\0\0\0\x08\0\0\0")?;

        let root_class = Class::Root {
            properties: self.0.clone(),
        };
        let buffer: Box<[u8]> = vec![0; root_class.rapified_length()].into_boxed_slice();
        let mut cursor = Cursor::new(buffer);
        let written = root_class.rapify(&mut cursor, 16)?;
        assert_eq!(written, root_class.rapified_length());

        let enum_offset = 16 + cursor.get_ref().len() as u32;
        output.write_u32::<LittleEndian>(enum_offset)?;

        output.write_all(cursor.get_ref())?;

        output.write_all(b"\0\0\0\0")?;
        assert_eq!(written + 20, self.rapified_length());
        Ok(written + 20)
    }

    fn rapified_length(&self) -> usize {
        let root_class = Class::Root {
            properties: self.0.clone(),
        };
        root_class.rapified_length() + 20 // metadata
    }
}
