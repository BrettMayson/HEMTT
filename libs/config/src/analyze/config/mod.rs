use std::sync::Arc;

use hemtt_common::config::ProjectConfig;
use hemtt_workspace::reporting::{Code, Processed};

use crate::Config;

use super::Analyze;

mod duplicate_classes;
mod duplicate_properties;
mod external_missing;
mod external_parent_case;
mod magwells;

impl Analyze for Config {
    fn warnings(
        &self,
        project: Option<&ProjectConfig>,
        processed: &Processed,
    ) -> Vec<Arc<dyn Code>> {
        let mut warnings = self
            .0
            .iter()
            .flat_map(|p| p.warnings(project, processed))
            .collect::<Vec<_>>();
        warnings.extend(external_parent_case::warn(&self.0, processed));
        if let Some(project) = project {
            warnings.extend(magwells::missing_magazine(project, self, processed));
        }
        warnings
    }

    fn errors(&self, project: Option<&ProjectConfig>, processed: &Processed) -> Vec<Arc<dyn Code>> {
        let mut errors = self
            .0
            .iter()
            .flat_map(|p| p.errors(project, processed))
            .collect::<Vec<_>>();
        errors.extend(duplicate_classes::error(&self.0, processed));
        errors.extend(external_missing::error(&self.0, processed));
        errors.extend(duplicate_properties::duplicate_properties(
            &self.0, processed,
        ));
        errors
    }
}
