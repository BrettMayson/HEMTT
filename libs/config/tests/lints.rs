#![allow(clippy::unwrap_used)]

use hemtt_common::config::ProjectConfig;
use hemtt_preprocessor::Processor;
use hemtt_workspace::{reporting::WorkspaceFiles, LayerType};

const ROOT: &str = "tests/lints/";

macro_rules! lint {
    ($dir:ident) => {
        paste::paste! {
            #[test]
            fn [<config_error_ $dir>]() {
                insta::assert_snapshot!(lint(stringify!($dir)));
            }
        }
    };
}

lint!(c01_invalid_value);
lint!(c01m_invalid_value_macro);
lint!(c02_duplicate_property_shadow_property);
lint!(c03_duplicate_class);
lint!(c03_duplicate_external);
lint!(c04_missing_parent);
lint!(c05_parent_case);
lint!(c06_unexpected_array);
lint!(c07_expected_array);
lint!(c08_missing_semicolon);
lint!(c09_magwell_missing_magazine);
lint!(c10_class_missing_braces);

fn lint(file: &str) -> String {
    let folder = std::path::PathBuf::from(ROOT);
    let workspace = hemtt_workspace::Workspace::builder()
        .physical(&folder, LayerType::Source)
        .finish(
            Some(ProjectConfig::test_project()),
            false,
            &hemtt_common::config::PDriveOption::Disallow,
        )
        .unwrap();
    let source = workspace.join(format!("{file}.hpp")).unwrap();
    let processed = Processor::run(&source).unwrap();
    let parsed = hemtt_config::parse(Some(&ProjectConfig::test_project()), &processed);
    let workspacefiles = WorkspaceFiles::new();
    match parsed {
        Ok(config) => config
            .codes()
            .iter()
            .map(|e| e.diagnostic().unwrap().to_string(&workspacefiles))
            .collect::<Vec<_>>()
            .join("\n")
            .replace('\r', ""),
        // Errors may occur, but they should be handled, if one is not a handler should be created
        Err(e) => {
            for e in &e {
                eprintln!("{}", e.diagnostic().unwrap().to_string(&workspacefiles));
            }
            panic!("Error parsing config");
        }
    }
}
