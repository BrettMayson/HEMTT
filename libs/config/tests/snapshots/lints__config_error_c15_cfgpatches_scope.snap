---
source: libs/config/tests/lints.rs
expression: lint(stringify! (c15_cfgpatches_scope)).0
---
[0m[1m[38;5;11mwarning[L-C15b][0m[1m: Public CfgVehicles class `abe_car2` not listed in CfgPatches[0m
   [0m[36mâ”Œâ”€[0m c15_cfgpatches_scope.hpp:13:11
   [0m[36mâ”‚[0m
[0m[36m13[0m [0m[36mâ”‚[0m     class [0m[33mabe_car2[0m: Car { // Missing
   [0m[36mâ”‚[0m           [0m[33m^^^^^^^^[0m [0m[33mhas scope=2 but not in units[][0m


[0m[1m[38;5;11mwarning[L-C15b][0m[1m: Public CfgVehicles class `abe_car5` not listed in CfgPatches[0m
   [0m[36mâ”Œâ”€[0m c15_cfgpatches_scope.hpp:22:11
   [0m[36mâ”‚[0m
[0m[36m22[0m [0m[36mâ”‚[0m     class [0m[33mabe_car5[0m: abe_car2 { }; // Missing
   [0m[36mâ”‚[0m           [0m[33m^^^^^^^^[0m [0m[33mhas scope=2 but not in units[][0m


[0m[1m[38;5;11mwarning[L-C15b][0m[1m: Public CfgWeapons class `abe_gun3` not listed in CfgPatches[0m
   [0m[36mâ”Œâ”€[0m c15_cfgpatches_scope.hpp:30:11
   [0m[36mâ”‚[0m
[0m[36m30[0m [0m[36mâ”‚[0m     class [0m[33mabe_gun3[0m: abe_gun2 {}; // Missing
   [0m[36mâ”‚[0m           [0m[33m^^^^^^^^[0m [0m[33mhas scope=2 but not in weapons[][0m


[0m[1m[38;5;11mwarning[L-C15a][0m[1m: CfgPatches units[] class `abe_bar22` not found[0m
  [0m[36mâ”Œâ”€[0m c15_cfgpatches_scope.hpp:3:44
  [0m[36mâ”‚[0m
[0m[36m3[0m [0m[36mâ”‚[0m         units[] = {"abe_car3", "abe_car4", [0m[33m"abe_bar22"[0m}; // bar sadly does not exist
  [0m[36mâ”‚[0m                                            [0m[33m^^^^^^^^^^^[0m [0m[33mnot defined in CfgVehicles[0m
